tasty-query/shared/src/main/scala/tastyquery/Trees.scala

```diff
@@ -102,11 +102,34 @@ object Trees {
   end StatementTree
 
   sealed abstract class TermTree(span: Span) extends StatementTree(span):
+    private var myTypePrefix: NonEmptyPrefix | Null = null
     private var myType: TermType | Null = null
 
     def withSpan(span: Span): TermTree
 
-    /** Compute the actual term of this tree. */
+    /** Compute the prefix represented by this tree.
+      *
+      * The default implementation throws an `InvalidProgramStructureException`
+      * to indicate that this term tree is not a valid type prefix.
+      */
+    protected def computeAsTypePrefix: NonEmptyPrefix =
```

Comment:
why this and not create a new PrefixTermTree?

Reply:
I considered that but it doesn't work. A `Select` has *sometimes* a valid `toTypePrefix`, sometimes not, depending on whether its `qualifier` has a valid `toTypePrefix` or not. There are similar problems with other kinds of trees. We would have needed a parallel hierarchy of these things, which wouldn't have played well with the unpickler anyway. So all in all, this is the only viable design.

Reply:
when would a qualifier not be a valid prefix?

Reply:
oh yeah, true

Reply:
so how does someone use this API, call it and hope there's no Apply in the prefix?

Reply:
Ok it seems to only be used to calculate the prefix of a singleton type tree, which should be properly constructed - int hat case would you consider making it `private[tastyquery]`?

Reply:
OK, I made it `private[Trees]`. And since it is then guaranteed to be called at most once per `TermTree`, I removed the cache field.